src/mpi/nsched/mpir_nsched.h

```diff
@@ -0,0 +1,97 @@
+/* -*- Mode: c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2016 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef MPIR_NSCHED_H_INCLUDED
+#define MPIR_NSCHED_H_INCLUDED
+
+#include "mpl_utlist.h"
+
+struct MPIR_Comm;
+
+typedef enum {
+    MPII_NSCHED_ELM_TYPE__SYNC,
+    MPII_NSCHED_ELM_TYPE__ISEND,
+    MPII_NSCHED_ELM_TYPE__IRECV
+} MPII_Nsched_elm_type_t;
+
+typedef struct MPII_Nsched_elm {
+    MPII_Nsched_elm_type_t type;
+
+    union {
+        struct {
+            const void *buf;
+            int count;
+            MPI_Datatype datatype;
+            int dest;
+            int tag;
+            struct MPIR_Comm *comm_ptr;
+        } isend;
+        struct {
+            void *buf;
+            int count;
+            MPI_Datatype datatype;
+            int dest;
+            int tag;
+            struct MPIR_Comm *comm_ptr;
+        } irecv;
+    } u;
+
+    int issued;
+    struct MPIR_Request *req;
+
+    struct MPII_Nsched_elm *next;
+} MPII_Nsched_elm;
+
+typedef struct MPIR_Nsched {
+    MPII_Nsched_elm *elm_list_head;
+    MPII_Nsched_elm *elm_list_tail;
+    struct MPIR_Request *req;
```

Comment:
This is the request for the entire collective and should remain active until the entire schedule is done.